contracts/src/RedemptionHelper.sol

```diff
@@ -0,0 +1,165 @@
+        }
+
+        // CS-BOLD-013: truncate redemption if it would exceed total unbacked debt
+        if (0 < totalProportions && totalProportions < _bold) _bold = totalProportions;
+
+        // Fallback: proportional to total debt
+        if (totalProportions == 0) {
+            for (uint256 i = 0; i < numBranches; ++i) {
+                branch[i].proportion = ITroveManager(branch[i].troveManager).getEntireBranchDebt();
+                if (branch[i].redeemable) totalProportions += branch[i].proportion;
+            }
+        }
+
+        if (totalProportions == 0) return (branch, totalProportions);
+
+        for (uint256 i = 0; i < numBranches; ++i) {
+            if (!branch[i].redeemable) continue;
+
+            branch[i].attemptedBold = _bold * branch[i].proportion / totalProportions;
+            if (branch[i].attemptedBold == 0) continue;
+
+            uint256 lastZombieTroveId = ITroveManager(branch[i].troveManager).lastZombieTroveId();
+            uint256 lastTroveId = ISortedTroves(branch[i].sortedTroves).getLast();
+
+            (uint256 troveId, uint256 nextTroveId) = lastZombieTroveId != 0
+                ? (lastZombieTroveId, lastTroveId)
+                : (lastTroveId, ISortedTroves(branch[i].sortedTroves).getPrev(lastTroveId));
+
+            for (
+                branch[i].iterations = 0;
+                branch[i].iterations < _maxIterationsPerCollateral || _maxIterationsPerCollateral == 0;
+                ++branch[i].iterations
+            ) {
+                if (branch[i].redeemedBold == branch[i].attemptedBold || troveId == 0) break;
+
+                LatestTroveData memory trove = ITroveManager(branch[i].troveManager).getLatestTroveData(troveId);
+                if (trove.entireColl * branch[i].price / trove.entireDebt >= _100pct) {
+                    branch[i].redeemedBold +=
+                        Math.min(branch[i].attemptedBold - branch[i].redeemedBold, trove.entireDebt);
+                }
+
+                troveId = nextTroveId;
+                nextTroveId = ISortedTroves(branch[i].sortedTroves).getPrev(nextTroveId);
+            }
+        }
+    }
+
+    // Meant to be called off-chain
+    // Not a view because price fetching has side-effects
+    function truncateRedemption(uint256 _bold, uint256 _maxIterationsPerCollateral)
+        external
+        returns (uint256 truncatedBold, uint256 feePct, Redeemed[] memory redeemed)
+    {
+        (SimulationContext[] memory branch, uint256 totalProportions) =
+            simulateRedemption(_bold, _maxIterationsPerCollateral);
+
+        if (totalProportions == 0) return (0, 0, redeemed);
+
+        truncatedBold = _bold;
+        for (uint256 i = 0; i < numBranches; ++i) {
+            if (branch[i].redeemable && branch[i].proportion > 0) {
+                uint256 extrapolatedBold = branch[i].redeemedBold * totalProportions / branch[i].proportion;
+                if (extrapolatedBold < truncatedBold) truncatedBold = extrapolatedBold;
```

Comment:
I’m having hard time understanding this part. It seems that `truncatedBold` can be overridden at any branch iteration. Why do we know the last one is the good one? I’ll get back to it later, but maybe a comment may help.